packages/flutter/lib/src/widgets/value_updater.dart

```diff
@@ -0,0 +1,171 @@
+
+  /// The value to pass to the [builder] when the widget is first built.
+  ///
+  /// Required.
+  final T initialValue;
+
+  /// The optional value to pass to the [builder] when the widget is built the
+  /// second and subsequent times.
+  ///
+  /// Ignored if [valueNotifier] is set.
+  ///
+  /// If not specified, then the current value of [valueNotifier] is used. One
+  /// of [value] or [valueNotifier] must be set.
+  ///
+  /// Only one of [valueNotifier] or [value] may be set at a time, but one of
+  /// them must be set.
+  final T? value;
+
+  /// The optional [ValueNotifier] to listen to.
+  ///
+  /// If specified, instead toggling between [initialValue] and [value] only
+  /// once, whenever the [ValueNotifier] notifies of a change, the widget will
+  /// use the notifier's value as the next value to build the [builder] with.
+  ///
+  /// Only one of [valueNotifier] or [value] may be set at a time, but one of
+  /// them must be set.
+  final ValueNotifier<T>? valueNotifier;
+
+  @override
+  State<ValueUpdater<T>> createState() => _ValueUpdaterState<T>();
+}
+
+enum _ValueUpdaterStatus {
+  initial,
+  off,
+  on,
+}
+
+class _ValueUpdaterState<T> extends State<ValueUpdater<T>> {
+  _ValueUpdaterStatus _triggered = _ValueUpdaterStatus.initial;
+  late T _value;
+
+  @override
+  void initState() {
+    super.initState();
+    _value = widget.value ?? widget.initialValue;
+    SchedulerBinding.instance.addPostFrameCallback((_) {
+      setState(() {
+        _value = widget.valueNotifier?.value ?? widget.value ?? _value;
+        _triggered = switch (_triggered) {
+          _ValueUpdaterStatus.off => _ValueUpdaterStatus.on,
+          _ValueUpdaterStatus.on => _ValueUpdaterStatus.off,
+          _ValueUpdaterStatus.initial => _ValueUpdaterStatus.on,
+        };
+      });
+    });
+    widget.valueNotifier?.addListener(_trigger);
+  }
+
+  @override
+  void dispose() {
+    widget.valueNotifier?.removeListener(_trigger);
+    super.dispose();
+  }
+
+  @override
+  void didUpdateWidget(covariant ValueUpdater<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.valueNotifier != widget.valueNotifier) {
+      oldWidget.valueNotifier?.removeListener(_trigger);
+      widget.valueNotifier?.addListener(_trigger);
+      _value = widget.valueNotifier?.value ?? widget.value ?? _value;
+    }
+    if (oldWidget.value != widget.value) {
+      final T? value = widget.value;
+      if (value != null) {
+        _value = value;
+      }
+    }
+  }
+
+  void _trigger() {
+    setState(() {
+      _value = widget.valueNotifier!.value;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return widget.builder(
+      context,
+      switch (_triggered) {
+        _ValueUpdaterStatus.off => widget.value != null ? widget.initialValue : _value,
+        _ValueUpdaterStatus.on => widget.value ?? _value,
+        _ValueUpdaterStatus.initial => widget.initialValue,
+      },
```

Comment:
Wondering: Do we even need to keep track of _triggered? If `_value` were to be assigned `widget.initialValue` in `initState` could we just always pass `_value` here and remove _ValueUpdaterStatus altogether?